policy-controller/k8s/index/src/http_route.rs

```diff
@@ -0,0 +1,136 @@
+use k8s_gateway_api::ParentReference;
+use linkerd_policy_controller_core::http_route::{
+    HeaderMatch, Hostname, HttpMethod, HttpRoute, HttpRouteMatch, PathMatch, QueryParamMatch, Value,
+};
+
+#[derive(Clone, Debug, PartialEq)]
+pub struct RouteBinding {
+    pub route: HttpRoute,
+    pub parent_refs: Vec<ParentReference>,
+}
+
+impl RouteBinding {
+    pub fn from_resource(route: k8s_gateway_api::HttpRoute) -> Self {
+        let hostnames = route
+            .spec
+            .hostnames
+            .iter()
+            .flatten()
+            .map(|hostname| {
+                if hostname.starts_with("*.") {
+                    let mut reverse_labels = hostname
+                        .split('.')
+                        .skip(1)
+                        .map(|label| label.to_owned())
+                        .collect::<Vec<String>>();
+                    reverse_labels.reverse();
+                    Hostname::Suffix { reverse_labels }
+                } else {
+                    Hostname::Exact(hostname.to_owned())
+                }
+            })
+            .collect();
+
+        let matches = route
+            .spec
+            .rules
+            .iter()
+            .flatten()
+            .filter_map(|rule| rule.matches.as_ref())
+            .flatten()
+            .map(|route_match| {
```

Comment:
Also, I'd generally suggest de-structuring `route_match` so that we can clearly illustrate we're consuming all of its fields. route_match could have fields that we don't use here, which would be undesirable. Also, i'd suggest using `into_iter` or cloning higher up, rather than having to clone/to_owned/etc on each field individually.